src/model/transform_modes.cu

```diff
@@ -190,150 +131,170 @@ void d_rotateForces(
 
 
 
-template<typename REAL>
+template<typename REAL, int PROTEINTYPE, bool MODES>
 void d_DOFPos(
 		unsigned blockSize,
 		unsigned gridSize,
 		const cudaStream_t &stream,
-		REAL const* xRec,
-		REAL const* yRec,
-		REAL const* zRec,
-		REAL const* xLig,
-		REAL const* yLig,
-		REAL const* zLig,
-		REAL const* xModesRec,
-		REAL const* yModesRec,
-		REAL const* zModesRec,
-		REAL const* xModesLig,
-		REAL const* yModesLig,
-		REAL const* zModesLig,
+		d_Protein<REAL>* protein,
 		DOF_6D_Modes<REAL>* dofs,
-		unsigned numAtomsRec,
-		unsigned numAtomsLig,
-		unsigned numModesRec,
-		unsigned numModesLig,
 		unsigned numDOFs,
-		REAL* xRecDefo,
-		REAL* yRecDefo,
-		REAL* zRecDefo,
-		REAL* xRecTrafo,
-		REAL* yRecTrafo,
-		REAL* zRecTrafo,
-		REAL* xLigDefo,
-		REAL* yLigDefo,
-		REAL* zLigDefo,
-		REAL* xLigTrafo,
-		REAL* yLigTrafo,
-		REAL* zLigTrafo)
-{
+		REAL* xDefo,
+		REAL* yDefo,
+		REAL* zDefo,
+		REAL* xTrafo,
+		REAL* yTrafo,
+		REAL* zTrafo
+		){
 	cudaVerifyKernel((
-			d_DOFPos<<<gridSize, blockSize, 0, stream>>> (
-			xRec,
-			yRec,
-			zRec,
-			xLig,
-			yLig,
-			zLig,
-			xModesRec,
-			yModesRec,
-			zModesRec,
-			xModesLig,
-			yModesLig,
-			zModesLig,
-			dofs,
-			numAtomsRec,
-			numAtomsLig,
-			numModesRec,
-			numModesLig,
-			numDOFs,
-			xRecDefo,
-			yRecDefo,
-			zRecDefo,
-			xRecTrafo,
-			yRecTrafo,
-			zRecTrafo,
-			xLigDefo,
-			yLigDefo,
-			zLigDefo,
-			xLigTrafo,
-			yLigTrafo,
-			zLigTrafo
+			d_DOFPos<REAL,PROTEINTYPE, MODES><<<gridSize, blockSize, 0, stream>>> (
+				protein->xPos,
+				protein->yPos,
+				protein->zPos,
+				protein->xModes,
+				protein->yModes,
+				protein->zModes,
+				dofs,
+				protein->numAtoms,
+				protein->numModes,
+				numDOFs,
+				xDefo,
+				yDefo,
+				zDefo,
+				xTrafo,
+				yTrafo,
+				zTrafo
 			))
 		);
 }
 
 template
-void d_DOFPos<float>(
+void d_DOFPos<float, 0, false >(
```

Comment:
I couldnt come up with a better solution. We need to differ weather we use modes or not. Otherwise unneeded calculation is done in the case, that modes are not used. And as long as we do only have dofs for the ligand, we need to change the dofs for the receptor. Or do you mean, that not all of the templates have to be declared?